connectors/sources/github.py

```diff
@@ -0,0 +1,252 @@
+                f"Error while filtering repositories. Exception: {exception}."
+            )
+            raise
+
+    @cached_property
+    def _get_session(self):
+        self.session = aiohttp.ClientSession(headers=self.headers)
+        return GitHubAPI(
+            self.session,
+            requester="",
+            oauth_token=self.github_token,
+        )
+
+    async def ping(self):
+        await self.execute_getitem(request=ENDPOINTS["USER"])
+
+    async def close_session(self):
+        self._sleeps.cancel()
+        if self.session is None:
+            return
+        await self.session.close()
+        del self._get_session
+
+
+class GitHubDataSource(BaseDataSource):
+    """GitHub"""
+
+    name = "GitHub"
+    service_type = "github"
+
+    def __init__(self, configuration):
+        """Setup the connection to the GitHub instance.
+
+        Args:
+            configuration (DataSourceConfiguration): Instance of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+        self._sleeps = CancellableSleeps()
+        self.github_client = GitHubClient(configuration=configuration)
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for GitHub.
+
+        Returns:
+            dict: Default configuration.
+        """
+        return {
+            "github_token": {
+                "label": "GitHub Token",
+                "sensitive": True,
+                "order": 1,
+                "type": "str",
+                "value": "changeme",
+            },
+            "repositories": {
+                "display": "textarea",
+                "label": "List of repositories",
+                "order": 2,
+                "type": "list",
+                "value": "*",
+            },
+            "retry_count": {
+                "display_value": RETRIES,
+                "display": "numeric",
+                "label": "Maximum retries per request",
+                "order": 3,
+                "required": False,
+                "type": "int",
+                "ui_restrictions": ["advanced"],
+                "value": RETRIES,
+                "validations": [{"type": "less_than", "constraint": 10}],
+            },
+        }
+
+    async def _remote_validation(self):
+        """Validate scope of the configured Token and accessibility of repositories
+
+        Raises:
+            ConfigurableFieldValueError: Insufficient privileges error.
+        """
+        response_headers = await self.github_client.execute_request(
+            method="get", url=USER_API
+        )
+        if "repo" not in response_headers.get("X-OAuth-Scopes"):  # pyright: ignore
+            raise ConfigurableFieldValueError(
+                "Configured token does not have required rights to fetch the content"
+            )
+        if self.github_client.repos != ["*"]:
+            invalid_repos = await self.github_client.filter_repos()
```

Comment:
suggestion: Maybe it makes sense to have a method fetching all public and private repositories for a user and do the check, if they match with the configured repositories (`list(set(configured_repos) - set(users_repos))`) in the data source and not in the client. To me it feels like only fetching the repositories should be the concern of the client and checking if it makes sense in terms of the configuration is a concern of the data source.

Reply:
Done